### ISSUE TITLE
'Failed to establish a new connection: [Errno 111] Connection refused' when trying to create a datacenter

### ISSUE THREAD

**Issue by vzovko** [ionos-cloud/module-ansible#1]
## Description
Playbook execution fails for datacenter creation with error:
```
'failed to set datacenter state: HTTPSConnectionPool(host=''api.ionos.com'', port=443): Max retries exceeded with url: /cloudapi/v5/datacenters?depth=2 (Caused by NewConnectionError(''<urllib3.connection.HTTPSConnection object at 0x7fe60b50a978>: Failed to establish a new connection: [Errno 111] Connection refused'',))'
```

## Expected behavior
Datacenter gets created using the Ionos Ansible module communicating over a proxy.

## Environment
Ansible version:
```
ansible 2.10.4
  config file = /home/<domain>/<username>/git/poc-ansible/playbooks/ionos/ansible.cfg
  configured module search path = ['/data/awx/venvs/modules/ionos_cloud']
  ansible python module location = /data/awx/venvs/ionos/lib/python3.6/site-packages/ansible
  executable location = /data/awx/venvs/ionos/bin/ansible
  python version = 3.6.9 (default, Oct  8 2020, 12:12:24) [GCC 8.4.0]
```
Module version: 
```
ionoscloud 5.0.3
module-ansible 5.0.1
```
OS: 
```
NAME="Ubuntu"
VERSION="18.04.5 LTS (Bionic Beaver)"
```

## Configuration Files
ansible.cfg
```
[defaults]
library = /data/awx/venvs/modules/ionos_cloud
host_key_checking = False
stdout_callback = yaml
```

Virtualenv setup for Playbook execution environment (prep.yml)
```yaml
- name: Create custom virutalenv for IONOS
  hosts: localhost
  gather_facts: no

  environment:
    HTTP_PROXY: http://proxy:port
    HTTPS_PROXY: http://proxy:port
  
  vars:
    venv_path: /data/awx/venvs/ionos
    module_path: /data/awx/venvs/modules/ionos_cloud

  tasks:
    - name: Create virutalenv and install packages from pip
      pip:
        name:
          - pip>=20.3.1
          - setuptools==51.0.0
          - wheel>=0.36.2
          - psutil==5.7.3
          - ansible>=2.10.4
          - jmespath>=0.10.0
          - cryptography>=3.3.1
          - ionoscloud>=5.0.3
        virtualenv: "{{ venv_path }}"
        virtualenv_command: /usr/bin/python3.6 -m venv
        virtualenv_site_packages: no
        umask: "0022"
      tags:
        - pip
        - base

    - name: Get Ionos Ansible Module
      git:
        repo: "https://github.com/ionos-cloud/module-ansible.git"
        dest: "{{ module_path }}"
      tags:
        - git
        - base
    
```

Test playbook (ionos.yml)
```yaml
- name: Demo
  hosts: localhost
  gather_facts: no
  
  environment:
    HTTP_PROXY: http://proxy:port
    HTTPS_PROXY: http://proxy:port
    IONOS_USERNAME: <vault encrypted>
    IONOS_PASSWORD: <vault encrypted>
  
  vars:
    datacenter_name: dc_test1
    datacenter_description: Test Datacenter 1
    datacenter_location: de/txl

  tasks:

    - name: Create datacenter
      datacenter:
        name: "{{ datacenter_name }}"
        description: "{{ datacenter_description }}"
        location: "{{ datacenter_location }}"
        state: present

```

## How to Reproduce
Create venv with above playbook `ansible-playbook prep.yml`

Activate venv with `source /data/awx/venvs/ionos/bin/activate`

Run playbook `ansible-playbook ionos.yml --vault-password-file ~/.ansible/vault.ionos`

## Error and Debug Output

```
> ansible-playbook ionos.yml --vault-password-file ~/.ansible/vault.ionos -vvv
ansible-playbook 2.10.4
  config file = /home/<domain>/<username>/git/poc-ansible/playbooks/ionos/ansible.cfg
  configured module search path = ['/data/awx/venvs/modules/ionos_cloud']
  ansible python module location = /data/awx/venvs/ionos/lib/python3.6/site-packages/ansible
  executable location = /data/awx/venvs/ionos/bin/ansible-playbook
  python version = 3.6.9 (default, Oct  8 2020, 12:12:24) [GCC 8.4.0]
Using /home/<domain>/<username>/git/poc-ansible/playbooks/ionos/ansible.cfg as config file
host_list declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
script declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Parsed /etc/ansible/hosts inventory source with ini plugin
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'
redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: ionos.yml ****************************************************************************************************************************************************************
1 plays in ionos.yml

PLAY [IONOS Demo Playbook] *********************************************************************************************************************************************************
META: ran handlers

TASK [Create datacenter] ***********************************************************************************************************************************************************
task path: /home/<domain>/<username>/git/poc-ansible/playbooks/ionos/ionos.yml:30
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: <domain>+<username>
<127.0.0.1> EXEC /bin/sh -c 'echo '"'"'~<domain>+<username>'"'"' && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo ~<domain>+<username>/.ansible/tmp `"&& mkdir "` echo ~<domain>+<username>/.ansible/tmp/ansible-tmp-1610546427.5803638-31746-29515677884723 `" && echo ansible-tmp-1610546427.5803638-31746-29515677884723="` echo ~<domain>+<username>/.ansible/tmp/ansible-tmp-1610546427.5803638-31746-29515677884723 `" ) && sleep 0'
Using module file /data/awx/venvs/modules/ionos_cloud/plugins/modules/datacenter.py
<127.0.0.1> PUT /home/<domain>/<username>/.ansible/tmp/ansible-local-3173897v1zgwt/tmp5h_6tm4r TO /home/<domain>/<username>/.ansible/tmp/ansible-tmp-1610546427.5803638-31746-29515677884723/AnsiballZ_datacenter.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /homes/<domain>/<username>/.ansible/tmp/ansible-tmp-1610546427.5803638-31746-29515677884723/ /homes/<domain>/<username>/.ansible/tmp/ansible-tmp-1610546427.5803638-31746-29515677884723/AnsiballZ_datacenter.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'HTTP_PROXY=http://proxy:port HTTPS_PROXY=http://proxy:port IONOS_USERNAME=<username@domain.de> IONOS_PASSWORD=<password> /data/awx/venvs/ionos/bin/python3.6 /homes/<domain>/<username>/.ansible/tmp/ansible-tmp-1610546427.5803638-31746-29515677884723/AnsiballZ_datacenter.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /homes/<domain>/<username>/.ansible/tmp/ansible-tmp-1610546427.5803638-31746-29515677884723/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/tmp/ansible_datacenter_payload_0q7glt0b/ansible_datacenter_payload.zip/ansible/modules/datacenter.py", line 376, in main
  File "/tmp/ansible_datacenter_payload_0q7glt0b/ansible_datacenter_payload.zip/ansible/modules/datacenter.py", line 189, in create_datacenter
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/ionoscloud/api/data_center_api.py", line 378, in datacenters_get
    return self.datacenters_get_with_http_info(**kwargs)  # noqa: E501
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/ionoscloud/api/data_center_api.py", line 510, in datacenters_get_with_http_info
    _request_auth=local_var_params.get('_request_auth'))
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/ionoscloud/api_client.py", line 423, in call_api
    _request_auth)
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/ionoscloud/api_client.py", line 226, in __call_api
    _request_timeout=_request_timeout)
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/ionoscloud/api_client.py", line 447, in request
    headers=headers)
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/ionoscloud/rest.py", line 234, in GET
    query_params=query_params)
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/ionoscloud/rest.py", line 212, in request
    headers=headers)
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/urllib3/request.py", line 75, in request
    method, url, fields=fields, headers=headers, **urlopen_kw
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/urllib3/request.py", line 96, in request_encode_url
    return self.urlopen(method, url, **extra_kw)
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/urllib3/poolmanager.py", line 375, in urlopen
    response = conn.urlopen(method, u.request_uri, **kw)
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/urllib3/connectionpool.py", line 796, in urlopen
    **response_kw
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/urllib3/connectionpool.py", line 796, in urlopen
    **response_kw
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/urllib3/connectionpool.py", line 796, in urlopen
    **response_kw
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/urllib3/connectionpool.py", line 756, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "/data/awx/venvs/ionos/lib/python3.6/site-packages/urllib3/util/retry.py", line 573, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
fatal: [localhost]: FAILED! => changed=false
  invocation:
    module_args:
      api_url: null
      description: Test Datacenter 1
      id: null
      location: de/txl
      name: dc_test1
      password: VALUE_SPECIFIED_IN_NO_LOG_PARAMETER
      state: present
      username: <username@domain.de>
      wait: true
      wait_timeout: 600
  msg: 'failed to set datacenter state: HTTPSConnectionPool(host=''api.ionos.com'', port=443): Max retries exceeded with url: /cloudapi/v5/datacenters?depth=2 (Caused by NewConnectionError(''<urllib3.connection.HTTPSConnection object at 0x7fe60b50a978>: Failed to establish a new connection: [Errno 111] Connection refused'',))'

PLAY RECAP *************************************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

```

## Additional Notes

Using the Ansible uri module to communicate directly with the API via proxy works just fine.

```yaml
- name: IONOS API
  hosts: localhost
  gather_facts: no

  environment:
    HTTP_PROXY: http://proxy:port
    HTTPS_PROXY: http://proxy:port
  
  vars:
    ionos_username: <vault encrypted>
    ionos_password: <vault encrypted>
    datacenter_name: dc_test1
    datacenter_description: Test Datacenter 1
    datacenter_location: de/txl

  tasks:
  
  - name: Create datacenter
    uri:
      url: https://api.ionos.com/cloudapi/v5/datacenters
      method: POST
      user: "{{ ionos_username }}"
      password: "{{ ionos_password }}"
      timeout: 60
      validate_certs: no
      return_content: yes
      headers:
        Accept: application/json
      body_format: json
      body:
        properties:
          name: "{{ datacenter_name }}"
          description: "{{ datacenter_description }}"
          location: "{{ datacenter_location }}"
      status_code: 202
    register: dc2
    tags:
      - create

  - debug:
      var: dc2
    tags:
      - create
```

**Result**
```
> ansible-playbook api.yml --vault-password-file ~/.ansible/vault.ionos --tags create
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [IONOS API] *******************************************************************************************************************************************************************

TASK [Create datacenter] ***********************************************************************************************************************************************************
ok: [localhost]

TASK [debug] ***********************************************************************************************************************************************************************
ok: [localhost] =>
  dc2:
    access_control_allow_credentials: 'true'
    access_control_allow_headers: '*'
    access_control_allow_methods: '*'
    access_control_allow_origin: '*'
    access_control_expose_headers: '*'
    changed: false
    connection: close
    content: |-
      {
        "id" : "863a8611-780b-4d25-87f7-18ea687b4fde",
        "type" : "datacenter",
        "href" : "https://api.ionos.com/cloudapi/v5/datacenters/863a8611-780b-4d25-87f7-18ea687b4fde",
        "metadata" : {
          "etag" : "c8ed9d884c20a380d29d0133efb3408b",
          "createdDate" : "2021-01-13T14:09:01Z",
          "createdBy" : "<username@domain.de>",
          "createdByUserId" : "df117326-9dbe-4acb-873d-77cb567d047d",
          "lastModifiedDate" : "2021-01-13T14:09:01Z",
          "lastModifiedBy" : "<username@domain.de>",
          "lastModifiedByUserId" : "df117326-9dbe-4acb-873d-77cb567d047d",
          "state" : "BUSY"
        },
        "properties" : {
          "name" : "dc_test1",
          "description" : "Test Datacenter 1",
          "location" : "de/txl",
          "version" : null,
          "features" : [ ],
          "secAuthProtection" : null
        }
      }
    content_length: '785'
    content_type: application/json
    cookies: {}
    cookies_string: ''
    date: Wed, 13 Jan 2021 14:09:01 GMT
    elapsed: 0
    etag: c8ed9d884c20a380d29d0133efb3408b
    failed: false
    json:
      href: https://api.ionos.com/cloudapi/v5/datacenters/863a8611-780b-4d25-87f7-18ea687b4fde
      id: 863a8611-780b-4d25-87f7-18ea687b4fde
      metadata:
        createdBy: <username@domain.de>
        createdByUserId: df117326-9dbe-4acb-873d-77cb567d047d
        createdDate: '2021-01-13T14:09:01Z'
        etag: c8ed9d884c20a380d29d0133efb3408b
        lastModifiedBy: <username@domain.de>
        lastModifiedByUserId: df117326-9dbe-4acb-873d-77cb567d047d
        lastModifiedDate: '2021-01-13T14:09:01Z'
        state: BUSY
      properties:
        description: Test Datacenter 1
        features: []
        location: de/txl
        name: dc_test1
        secAuthProtection: null
        version: null
      type: datacenter
    location: https://api.ionos.com/cloudapi/v5/requests/ef5e9020-40c6-424d-8ade-a7e1745da982/status
    msg: OK (785 bytes)
    redirected: false
    server: Apache/2.4.38 (Debian)
    status: 202
    url: https://api.ionos.com/cloudapi/v5/datacenters
    x_ratelimit_burst: '50'
    x_ratelimit_limit: '120'
    x_ratelimit_remaining: '49'

```

## References

https://github.com/ionos-cloud/module-ansible#datacenter

**Comment by vzovko** [comment]
After digging around I noticed that in the Python SDK in file https://github.com/ionos-cloud/sdk-python/blob/335b1706052505ddbcb466345522a077887272ab/ionoscloud/configuration.py#L215 proxy gets set to None.

I modified the file an added `import os` at the top and modified the line 215 to `self.proxy = os.environ.get('HTTP_PROXY')`. Now it is working for me.

Is this something which can be resolved permanently in future Python SDK versions?
